In [ ]:
 !pip install transformers
!pip install datasets

     |████████████████████████████████| 2.2MB 17.6MB/s 
     |████████████████████████████████| 3.3MB 52.9MB/s 
     |████████████████████████████████| 870kB 28.4MB/s 
  Created wheel for sacremoses: filename=sacremoses-0.0.44-cp37-none-any.whl size=886084 sha256=d3062787088d43700cf047c8075314f78c6134eee13b9133be63328e9e8c5155
  Stored in directory: /root/.cache/pip/wheels/3e/fb/c0/13ab4d63d537658f448366744654323077c4d90069b6512f3c
Successfully built sacremoses
     |████████████████████████████████| 194kB 17.2MB/s 
     |████████████████████████████████| 112kB 37.9MB/s 
     |████████████████████████████████| 245kB 49.4MB/s 


In [ ]:
from transformers import AutoTokenizer, AutoModelWithLMHead
  
tokenizer = AutoTokenizer.from_pretrained("t5-base")
  
model = AutoModelWithLMHead.from_pretrained("t5-base")

In [ ]:
model.cuda()

In [ ]:
"""
Hans is nice. premise: Hans is immortal. wird zu entailment

Hans is nice. premise: Hans is mortal wird zu contradiction.

(umgedreht wenn ugly statt nice)

T5 schaut also nicht wirklich auf logik, sondern ob beide adjektive negativ/positiv annotiert sind
"""

In [ ]:
"""
T5 schaut auch auf token overlapps

hypothesis: lgisbs qk  asdof asdsadf sdaf opekewqtj+ öysdohidfgölmn premise: df opiasejiü nüowe mjdbf0e <,m opekewqtj+ öysdohidfgölmn sfio´wetmknyxgbkjbbu 
wird zu entailment während

hypothesis: lgisbs qk  asdof asdsadf sdaf opekewqtj+ öysdohidfgölmn premise: df opiasejiü nüowe mjdbf0e <,m sfio´wetmknyxgbkjbbu

contradiction ist

(im ersten ist der prämisse ein teil der hypothese eingefügt)

"""

In [ ]:
"hypothesis: A person loves gras. premise: All humans like turf" wird zu contradiciton
text = "hypothesis: Trees are green premise: Plants are green"wird zu neutral-> versteht also nicht dass bäume auch plzanzen sind

In [ ]:
premise = tokenizer.encode("All humans are mortal. Hans is a human. ",return_tensors="pt").cuda()[:,1:]
hypothesis = tokenizer.encode("Hans is mortal. ",return_tensors="pt").cuda()[:,1:-1]
X = tokenizer.encode("hypothesis: ",return_tensors="pt").cuda()[:,:-1]
Y = tokenizer.encode("premise: ",return_tensors="pt").cuda()[:,1:-1]


input_ids = torch.cat([X,hypothesis,Y,premise],dim=-1)
print(input_ids)
print(tokenizer.encode("hypothesis: Hans is mortal. premise: All humans are mortal. Hans is a human. ",return_tensors="pt").cuda())
print(tokenizer.decode(input_ids[0]))

tokenizer.decode(model.generate(input_ids)[0])

tensor([[22455,    10,    19, 24301,     5, 17398,    10,  6917,    33, 24301,
             5, 10001,    19,     3,     9,   936,     5,     1]],
       device='cuda:0')
tensor([[22455,    10, 10001,    19, 24301,     5,     3, 17398,    10,   432,
          6917,    33, 24301,     5, 10001,    19,     3,     9,   936,     5,
             1]], device='cuda:0')
hypothesis: is mortal.premise: humans are mortal. Hans is a human.</s>


'<pad> entailment</s>'

In [ ]:


print(input_ids)
# With view
idx = torch.randperm(input_ids.nelement())
tokenizer.decode(input_ids.view(-1)[idx].view(input_ids.size())[0])

tensor([[10001,    19,     3,     9,   936,     5,   432,  6917,    33, 24301,
             5,     1]], device='cuda:0')


'humansa. are is human.</s> All mortal Hans '

In [ ]:
: mnli hypothesis: The St. Louis Cardinals have always won. premise:
yeah well losing is i mean i’m i’m originally from Saint Louis and Saint Louis
Cardinals when they were there were uh a mostly a losing team but


# Translate

In [ ]:
pip install googletrans==3.1.0a0

     |████████████████████████████████| 61kB 5.4MB/s 
     |████████████████████████████████| 1.0MB 16.8MB/s 
     |████████████████████████████████| 51kB 6.1MB/s 
     |████████████████████████████████| 61kB 7.0MB/s 
     |████████████████████████████████| 71kB 7.1MB/s 
  Created wheel for googletrans: filename=googletrans-3.1.0a0-cp37-none-any.whl size=16368 sha256=de81d8a2db1a7dd88ce398209cc50f4bcff3d3e6402f6b96dce273a15c847943
  Stored in directory: /root/.cache/pip/wheels/27/7a/a0/aff3babbb775549ce6813cb8fa7ff3c0848c4dc62c20f8fdac
Successfully built googletrans


In [ ]:
#TODO do again, minimize API access, or use another transformer(MartianMT), then do other languages like spanish
from google_trans_new import google_translator  
translator = google_translator()  
translate_text = translator.translate('I like the stock',lang_tgt='de',lang_src="en")  
print(translate_text)
#output: Hello china

ModuleNotFoundError: ignored

In [ ]:
!pip install translate

     |████████████████████████████████| 194kB 10.2MB/s 
     |████████████████████████████████| 92kB 8.3MB/s 
     |████████████████████████████████| 7.2MB 16.3MB/s 
     |████████████████████████████████| 645kB 50.9MB/s 
     |████████████████████████████████| 102kB 13.4MB/s 
     |████████████████████████████████| 337kB 50.2MB/s 
ERROR: pytest 3.6.4 has requirement pluggy<0.8,>=0.5, but you'll have pluggy 0.13.1 which is incompatible.
ERROR: datascience 0.10.6 has requirement folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
  Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13
  Found existing installation: pluggy 0.7.1
    Uninstalling pluggy-0.7.1:
      Successfully uninstalled pluggy-0.7.1


In [ ]:
 import googletrans
 googletrans.LANGCODES

In [ ]:
[]

ValueError: ignored

In [ ]:
!pip install dl-translate

In [ ]:
!pip install sentencepiece

In [ ]:
import dl_translate as dlt
mt = dlt.TranslationModel()

In [ ]:
text_hi = "संयुक्त राष्ट्र के प्रमुख का कहना है कि सीरिया में कोई सैन्य समाधान नहीं है"
mt.translate(text_hi, source=dlt.lang.HINDI, target=dlt.lang.ENGLISH)

'The head of the United Nations says there is no military solution in Syria'

In [ ]:
text = "I like the stock"
mt.translate(text, source=dlt.lang.ENGLISH, target=dlt.lang.SWEDISH)

'Jag gillar aktien'

# Dataset

In [ ]:
from datasets import load_dataset
import torch
dataset = load_dataset("multi_nli",split="validation_matched[:400]")
batch_size = 1
dataloader_mnli = torch.utils.data.DataLoader(dataset, batch_size=batch_size,num_workers=2,pin_memory=True)
mnli_length = len(dataset)
mnli_length

Using custom data configuration default


Dataset multi_nli downloaded and prepared to /root/.cache/huggingface/datasets/multi_nli/default/0.0.0/3248359997b13e6ccd296f42420b31c107ba6859b742ed6af1dce0f1544c9ec1. Subsequent calls will reuse this data.


400

In [ ]:
prompts = [["mnli ",""],
           ["premise: ", "Consider the following premise: ", "The following premise is true: ", " "],
           [" hypothesis: ", " Consider the following hypothesis: ", " The following hypothesis is true: ", " "],
           ["", " The label is"]]

prompts_length = len(prompts)

In [ ]:
prompts = [["","This is a task that requires solid reasoning, clear thinking and rigorous inference. "],
           ["We will assume that the following statement is true: ", "I believe that the following statement is true: ", "Yellow is evil. ","bark train can tee"," "],
           [" Furthermore, we will assume that the next statement is true aswell: ", "Furthermore, I believe that the next statement is true: ","Apples can fly. ", "brain fly science ananas"," "],
           ["", "With this in mind, we can conclue that the label should be"]]

prompts_length = len(prompts)

In [ ]:
prompts = [[""],
           ["Consider the following hypothesis: "],
           [" premise: "],
           [""]]
prompts_length = len(prompts)

In [ ]:
prompts = [[""],
           ["hypothesis: ","Consider the following hypothesis: "],
           [" premise: "],
           [""]]
prompts_length = len(prompts)

In [ ]:
dataset[58]

{'genre': 'slate',
 'hypothesis': 'She has chose to live a hollow life.',
 'hypothesis_binary_parse': '( She ( ( has ( chose ( to ( live ( a ( hollow life ) ) ) ) ) ) . ) )',
 'hypothesis_parse': '(ROOT (S (NP (PRP She)) (VP (VBZ has) (VP (VBD chose) (S (VP (TO to) (VP (VB live) (NP (DT a) (JJ hollow) (NN life))))))) (. .)))',
 'label': 2,
 'pairID': '13316c',
 'premise': 'She has exchanged a hollow life for a heightened life, and has tried to comprehend all its turns, get its possibilities.',
 'premise_binary_parse': '( She ( ( ( ( ( has ( exchanged ( ( a ( hollow life ) ) ( for ( a ( heightened life ) ) ) ) ) ) , ) and ) ( has ( tried ( to ( ( ( comprehend ( all ( its turns ) ) ) , ) ( get ( its possibilities ) ) ) ) ) ) ) . ) )',
 'premise_parse': '(ROOT (S (NP (PRP She)) (VP (VP (VBZ has) (VP (VBN exchanged) (NP (NP (DT a) (JJ hollow) (NN life)) (PP (IN for) (NP (DT a) (VBN heightened) (NN life)))))) (, ,) (CC and) (VP (VBZ has) (VP (VBN tried) (S (VP (TO to) (VP (VP (VB comprehend

In [ ]:
from tqdm import tqdm_notebook
#TDOD change again to premise, hypothesis instead of the other way around
accuracies = dict.fromkeys(prompts[0] , 0)
accuracies = {x: dict.fromkeys(prompts[1] , 0) for x in accuracies}
for key,value in accuracies.items():
  accuracies[key] =  {x: dict.fromkeys(prompts[2] , 0) for x in value}
for key,value in accuracies.items():
  for key2,value2 in value.items():
    accuracies[key][key2] =  {x: dict.fromkeys(prompts[3] , 0) for x in value2}

len_ent = 0
len_neu = 0
len_contr = 0

counter =0 

for i, mnli_instance in enumerate(tqdm_notebook(dataloader_mnli)):
      for q,trigger in enumerate(prompts[0]):
        for j,X in enumerate(prompts[1]):
          for k,Y in enumerate(prompts[2]):
            for l,Z in enumerate(prompts[3]):
              hypothesis = mnli_instance["hypothesis"][0]
              
              #hypothesis = list(hypothesis)
              #hypothesis = ["i" if x=="e"  else x  for x in hypothesis]
              #hypothesis = ["e" if x=="a"  else x  for x in hypothesis]
              #hypothesis = ["a" if x=="o"  else x  for x in hypothesis]
              #hypothesis = ["o" if x=="u"  else x  for x in hypothesis]
              #hypothesis = "".join(hypothesis)
              #hypothesis = mt.translate(mnli_instance["hypothesis"][0], source=dlt.lang.ENGLISH, target=dlt.lang.SWEDISH)
              #hypothesis = translator.translate(mnli_instance["hypothesis"][0],lang_tgt='de',lang_src="en")
              #input_ids = tokenizer.encode(text,return_tensors="pt").cuda()
              #hypothesis = tokenizer.decode(model.generate(input_ids,num_beams=3)[0],skip_special_tokens=True)
              

              premise = mnli_instance["premise"][0]
              #if len(hypothesis)/len(premise)  > 0.1:
              #  continue
              #premise = list(premise)
              #premise = ["i" if x=="e"  else x  for x in premise]
              #premise = ["e" if x=="a"  else x  for x in premise]
              #premise = ["a" if x=="o"  else x  for x in premise]
              #premise = ["o" if x=="u"  else x  for x in premise]
              #premise = "".join(premise)

              #premise = mt.translate(mnli_instance["premise"][0], source=dlt.lang.ENGLISH, target=dlt.lang.SWEDISH)
              #premise = translator.translate(mnli_instance["premise"][0],lang_tgt='de',lang_src="en")
              #text = "translate English to Romanian: " + mnli_instance["premise"][0]
              #input_ids = tokenizer.encode(text,return_tensors="pt").cuda()
              #premise = tokenizer.decode(model.generate(input_ids,num_beams=3)[0],skip_special_tokens=True)
              
              
              counter+=1
              text = trigger + X + premise + Y + hypothesis + Z
              input_ids = tokenizer.encode(text,return_tensors="pt").cuda()
              label = tokenizer.decode(model.generate(input_ids,num_beams=3)[0],skip_special_tokens=True)
              if label == "entailment":
                len_ent +=len(hypothesis)/len(premise) 
                label = 0
              elif label == "neutral":
                len_neu+= len(hypothesis)/len(premise) 
                label = 1
              elif label == "contradiction":
                len_contr+=len(hypothesis)/len(premise) 
                label = 2

              if label == mnli_instance["label"] :#or (label==1 and mnli_instance["label"]==0)
                accuracies[trigger][X][Y][Z]+=1
              #else:
              #  print(text)
              #  print(label)
              #  print(mnli_instance["label"])
              #  print()


accuracies

#dict statt list

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:17: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`


{'': {'Consider the following hypothesis: ': {' premise: ': {'': 220}}}}

In [ ]:
print(len_ent/counter)
print(len_neu/counter)
print(len_contr/counter)

0.0
0.41877066608358887
0.24863980975965547


In [ ]:
import copy
accuracies_p = copy.deepcopy(accuracies)

for key0,value0 in accuracies_p.items():
  for key1,value1 in value0.items() :
    for key2,value2 in value1.items() :
      for key3,value3 in value2.items():
        accuracies_p[key0][key1][key2][key3] = value3/counter#mnli_len

accuracies_p



{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.55}}}}

In [ ]:
# mnli validation matched[0:400]
#premise und hypothesis vertauscht, immer noch acc weit über random
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.55}}}}

In [ ]:
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.7725}}}}#cats are green
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.8475}}}}#"It is midly inconvient to book a train ticket and end on a ship. Furthermore, "
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.8225}}}}#"A is bad,terrible and nasty. Furthermore, "

In [ ]:
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.855}}}}
# bei "James ate a cake and " kein acc drop
#t5 gibt hier tatsächlich nie "entailment aus"

In [ ]:
# mnli validation matched[0:400]
# wenn man zur hypothese noch etwas hinzufügt, was sich aus der prämisse nicht ableiten lassen kann("All humans are basketballs and "),
#dann sollte das label neutral sein für mnli_label neutral und entailment und contradiciton für mnli_label contradiction
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.7625}}}}
# also immer noch weit über rand, aber acc drop von 10%

In [ ]:
len(hypothesis)/len(premise) ist bei ent höhher wie bei premise
# wenn man nur examples nimmt, bei denen dieser wert niedrig ist(0.1), so sinkt acc auf 80%
#kann aber auch sein, dass dies an anzahl der trainingsexample liegt die übrig bleibemn(46), bei gesammten testset

In [ ]:
#evtl höhere acc wenn man mehr testexample nutzt
0.867647058823529 bei8000, aber nur welche mit len premise > 30

In [ ]:
36.7175
37.525
39.285

len premise

In [ ]:
#keine unterschiede in der länge der hyp zwischen ent,neu,cont also hier kein statistical cue

In [ ]:
# mnli validation matched[0:200]
#alle vocale durch andere ersetzten, hier immer noch acc leicht über random. Kann aber sein, dass Menschen da auch noch einige richtig raten können
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.47}},
  'hypothesis: ': {' premise: ': {'': 0.425}}}}

In [ ]:
# mnli validation matched[0:200]
# alle "e" durch "i" ersetzen
# acc weit über random, ist natürlich fraglich inwieweit dies menschen auch können würden
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.675}},
  'hypothesis: ': {' premise: ': {'': 0.695}}}}

In [ ]:
# mnli validation matched[0:200]
# random text generally degrades acc
{'': {'I like the following hypothesis: ': {' premise: ': {'': 0.655}},
  'May the force be with this hypothesis: ': {' premise: ': {'': 0.515}},
  'believe the following hypothesis: ': {' premise: ': {'': 0.81}},
  'hypothesis: ': {' premise: ': {'': 0.83}}}}

In [ ]:
#mnli validation matched[0:200]
# zu viel text und auch kurzer random text lassen die acc sehr sinken
{'': {'Please use logical thinking, be rational and carefully consider all the alternatives. hypothesis: ': {' premise: ': {'': 0.53}},
  'apple has tree of piece hypothesis:': {' premise: ': {'': 0.51}},
  'consider the following hypothesis:': {' premise: ': {'': 0.835}},
  'hypothesis: ': {' premise: ': {'': 0.83}}}}

In [ ]:
#mnli validation matched[0:200]
#casting it to SQUAD format, only works if we still have mnli triggers in the context
#acc degrades compared to normal mnli, but still better than random
{'question: Should the label be entailment, contradiction or neutral? context: ': {'hypothesis: ': {' premise: ': {'': 0.5}}}}

In [ ]:
# bei copa ist nicht klar wie man casten soll, es kommt eine mischung von mnli,copa udn rte labeln raus

In [ ]:
#mnli validation matched[0:200]
#qnli bekommt eine frage und einen Kontext. Die Aufgabe ist es zu bestimmen, ob man die Frage aus dem vorliegenden text beantworten kann.
#hier einfach mal hypothese als question und premise als kontext genommen -  bringt acc über random, obwohl nie "neutral ausgegeben werden kann"(also maximal errreichbare acc bei 66% oder so)
{'': {'question: ': {' sentence: ': {'': 0.04}}},
 'qnli ': {'question: ': {' sentence: ': {'': 0.515}}}}

In [ ]:
#mnli validation matched[0:200]
#selbes mrpc experiment, aber premise und hyp geswapped
# hat niedrigere acc, was in diesem Fall ja besser ist,aber komisch, dass es bei Äquivalenz auf die reihenfolge ankommt
{'mrpc ': {'sentence1: ': {' sentence2: ': {'': 0.435}}}}

In [ ]:
#mnli validation matched[0:200]
#mrpc task ist entscheidung ob sätze äquivalent sind oder nicht
#hier "equivalent" = "entailment" und "not_equivalent" = "contradiction"
#tatsächlich etwas über random acc
{'mrpc ': {'sentence1: ': {' sentence2: ': {'': 0.47}}}}

In [ ]:
#mnli validation matched[0:200]
#using mBART large to translate both premise and hypothesis to swedish
# etwas mehr wie random acc, aber nicht viel
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.405}},
  'hypothesis: ': {' premise: ': {'': 0.445}}}}

In [ ]:
#mnli validation matched[0:200]
#using mBART large to translate both premise and hypothesis to korean
# hier geht es auf random acc runter, was zu erwarten war...
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.41}},
  'hypothesis: ': {' premise: ': {'': 0.415}}}}

In [ ]:
#mnli validation matched[0:200]
#using mBART large to translate both premise and hypothesis to spanish
# still significantly above random acc even though it was never finetuned to translate to spanish?
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.59}},
  'hypothesis: ': {' premise: ': {'': 0.58}}}}

In [ ]:
#mnli validation matched[0:200]
#using mBART large to translate both premise and hypothesis to german
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.69}},
  'hypothesis: ': {' premise: ': {'': 0.695}}}}

In [ ]:
#mnli validation matched[0:200]
#zwei mal premise, hier ist acc gleich wie zwei mal hypothesis, also hat nur entailemnt predicten auf diesem testsubset 41% acc
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.41}},
  'hypothesis: ': {' premise: ': {'': 0.41}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.405}},
  'hypothesis: ': {' premise: ': {'': 0.41}}}}

In [ ]:
#mnli validation matched[0:200]
#zwei mal hypothese auch leicht über rand chance(sollte eig nur entailment ausgeben?)
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.405}},
  'hypothesis: ': {' premise: ': {'': 0.41}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.41}},
  'hypothesis: ': {' premise: ': {'': 0.41}}}}

In [ ]:
#mnli validation matched[0:200]
#nur die hypothese auch acc leicht über random chance
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.375}},
  'hypothesis: ': {' premise: ': {'': 0.375}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.375}},
  'hypothesis: ': {' premise: ': {'': 0.395}}}}

In [ ]:
#mnli validation matched[0:200]
#nur die premise führt zu acc leicht über random chance
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.4}},
  'hypothesis: ': {' premise: ': {'': 0.385}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.405}},
  'hypothesis: ': {' premise: ': {'': 0.39}}}}

In [ ]:
#mnli validation matched[0:300]
#o cb oder mnli oder ganz weglassen scheint keinen großen Unterschied zu machen
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.845}},
  'hypothesis: ': {' premise: ': {'': 0.83}}},
 'cb ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.835}},
  'hypothesis: ': {' premise: ': {'': 0.835}}}}

In [ ]:
#mnli validation matched[0:300], google translate to hypothesis and premise to german seperatly, then do standard mnli,beam=3
#immer noch acc weit über 36% random acc
#nvm, translation just didnt work
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.8366666666666667}},
  'hypothesis: ': {' premise: ': {'': 0.8366666666666667}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.8266666666666667}},
  'hypothesis: ': {' premise: ': {'': 0.8233333333333334}}}}

In [ ]:
#mnli validation matched[0:300], let t5 first translate hypothesis and premise seperatly to romanian, then do standard mnli,beam=3
#immer noch acc weit über 36% random acc
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.5466666666666666}},
  'hypothesis: ': {' premise: ': {'': 0.5466666666666666}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.57}},
  'hypothesis: ': {' premise: ': {'': 0.5466666666666666}}}}

In [ ]:
#mnli validation matched[0:300], let t5 first translate hypothesis and premise seperatly to french, then do standard mnli,beam=3
#immer noch acc weit über 36% random acc
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.5433333333333333}},
  'hypothesis: ': {' premise: ': {'': 0.5366666666666666}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.5433333333333333}},
  'hypothesis: ': {' premise: ': {'': 0.5433333333333333}}}}

In [ ]:
#mnli validation matched[0:300], let t5 first translate hypothesis and premise seperatly to german, then do standard mnli,beam=3
#immer noch acc weit über 36% random acc
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.6033333333333334}},
  'hypothesis: ': {' premise: ': {'': 0.6}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.6133333333333333}},
  'hypothesis: ': {' premise: ': {'': 0.5833333333333334}}}}

In [ ]:
#mnli validation matched[0:300], let t5 first summarize the hypothesis, then do standard mnli,beam=3
#erstaunlicherweise acc gains von bis zu 3%(kann an 300 statt 200 liegen oder auch daran dass t5 satz so anpasst,dass es für mnli passt)
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.8566666666666667}},
  'hypothesis: ': {' premise: ': {'': 0.8566666666666667}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.8466666666666667}},
  'hypothesis: ': {' premise: ': {'': 0.8533333333333334}}}}

In [ ]:
# snli,0:300 test, premise and hypothesis swapped,beam search mit num_beams=6
{'': {'Consider the following hypothesis: ': {' premise: ': {'': 0.8166666666666667}},
  'hypothesis: ': {' premise: ': {'': 0.8066666666666666}}},
 'mnli ': {'Consider the following hypothesis: ': {' premise: ': {'': 0.7866666666666666}},
  'hypothesis: ': {' premise: ': {'': 0.78}}}}

In [ ]:
# mnli,0:200 validation matched, premise and hypothesis swapped,beamm search mit num_beams=6
# wie erwarted gleiche acc wie b=3, denn 3 beams reichen aus um sich neutral,contradiction und entailment zu merken
{'': {'Consider the following hypothesis: ': {' Consider the following premise: ': {'': 0.775},
   ' premise: ': {'': 0.845}},
  'hypothesis: ': {' Consider the following premise: ': {'': 0.84},
   ' premise: ': {'': 0.83}}},
 'mnli ': {'Consider the following hypothesis: ': {' Consider the following premise: ': {'': 0.76},
   ' premise: ': {'': 0.83}},
  'hypothesis: ': {' Consider the following premise: ': {'': 0.81},
   ' premise: ': {'': 0.825}}}}

In [ ]:
# mnli,0:200 validation matched, premise and hypothesis swapped,beamm search mit num_beams=3
# erhöht acc gegenüber greedy etwas? vlt weil es ein paar wenige example gibt, bei denen nicht eines der drei label rauskommt als erster token...
{'': {'Consider the following hypothesis: ': {' Consider the following premise: ': {'': 0.775},
   ' premise: ': {'': 0.845}},
  'hypothesis: ': {' Consider the following premise: ': {'': 0.84},
   ' premise: ': {'': 0.83}}},
 'mnli ': {'Consider the following hypothesis: ': {' Consider the following premise: ': {'': 0.76},
   ' premise: ': {'': 0.83}},
  'hypothesis: ': {' Consider the following premise: ': {'': 0.81},
   ' premise: ': {'': 0.825}}}}

In [ ]:
# mnli,0:200 validation matched, premise and hypothesis swapped
#sampling=True,lower acc as expected
{'': {'Consider the following hypothesis: ': {' Consider the following premise: ': {'': 0.69},
   ' premise: ': {'': 0.75}},
  'hypothesis: ': {' Consider the following premise: ': {'': 0.785},
   ' premise: ': {'': 0.815}}},
 'mnli ': {'Consider the following hypothesis: ': {' Consider the following premise: ': {'': 0.685},
   ' premise: ': {'': 0.73}},
  'hypothesis: ': {' Consider the following premise: ': {'': 0.72},
   ' premise: ': {'': 0.78}}}}

In [ ]:
# mnli,0:200 validation matched, premise and hypothesis swapped
# expected based on the paper is 87% with fine-tuning
{'': {'Consider the following hypothesis: ': {' Consider the following premise: ': {'': 0.72},
   ' premise: ': {'': 0.83}},
  'hypothesis: ': {' Consider the following premise: ': {'': 0.825},
   ' premise: ': {'': 0.825}}},
 'mnli ': {'Consider the following hypothesis: ': {' Consider the following premise: ': {'': 0.72},
   ' premise: ': {'': 0.83}},
  'hypothesis: ': {' Consider the following premise: ': {'': 0.785},
   ' premise: ': {'': 0.805}}}}

In [ ]:
# mnli,0:500 validation matched

{'': {' ': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}},
  'I believe that the following statement is true: ': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}},
  'We will assume that the following statement is true: ': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}},
  'Yellow is evil. ': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}},
  'bark train can tee': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}}},
 'This is a task that requires solid reasoning, clear thinking and rigorous inference. ': {' ': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}},
  'I believe that the following statement is true: ': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}},
  'We will assume that the following statement is true: ': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}},
  'Yellow is evil. ': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}},
  'bark train can tee': {' ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   ' Furthermore, we will assume that the next statement is true aswell: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Apples can fly. ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'Furthermore, I believe that the next statement is true: ': {'': 0,
    'With this in mind, we can conclue that the label should be': 0},
   'brain fly science ananas': {'': 0,
    'With this in mind, we can conclue that the label should be': 0}}}}

In [ ]:
# mnli,500:1000 validation matched

{'': {' ': {' ': {'': 0.0, ' The label is': 0.0},
   ' Consider the following hypothesis: ': {'': 0.614, ' The label is': 0.606},
   ' The following hypothesis is true: ': {'': 0.164, ' The label is': 0.148},
   ' hypothesis: ': {'': 0.604, ' The label is': 0.604}},
  'Consider the following premise: ': {' ': {'': 0.416,
    ' The label is': 0.376},
   ' Consider the following hypothesis: ': {'': 0.562, ' The label is': 0.552},
   ' The following hypothesis is true: ': {'': 0.526, ' The label is': 0.536},
   ' hypothesis: ': {'': 0.594, ' The label is': 0.59}},
  'The following premise is true: ': {' ': {'': 0.166, ' The label is': 0.136},
   ' Consider the following hypothesis: ': {'': 0.632, ' The label is': 0.598},
   ' The following hypothesis is true: ': {'': 0.506, ' The label is': 0.468},
   ' hypothesis: ': {'': 0.606, ' The label is': 0.59}},
  'premise: ': {' ': {'': 0.514, ' The label is': 0.522},
   ' Consider the following hypothesis: ': {'': 0.59, ' The label is': 0.59},
   ' The following hypothesis is true: ': {'': 0.592, ' The label is': 0.594},
   ' hypothesis: ': {'': 0.596, ' The label is': 0.586}}},
 'mnli ': {' ': {' ': {'': 0.0, ' The label is': 0.0},
   ' Consider the following hypothesis: ': {'': 0.616, ' The label is': 0.612},
   ' The following hypothesis is true: ': {'': 0.396, ' The label is': 0.386},
   ' hypothesis: ': {'': 0.59, ' The label is': 0.584}},
  'Consider the following premise: ': {' ': {'': 0.442,
    ' The label is': 0.416},
   ' Consider the following hypothesis: ': {'': 0.588, ' The label is': 0.59},
   ' The following hypothesis is true: ': {'': 0.562, ' The label is': 0.568},
   ' hypothesis: ': {'': 0.578, ' The label is': 0.582}},
  'The following premise is true: ': {' ': {'': 0.374, ' The label is': 0.364},
   ' Consider the following hypothesis: ': {'': 0.552, ' The label is': 0.546},
   ' The following hypothesis is true: ': {'': 0.558, ' The label is': 0.55},
   ' hypothesis: ': {'': 0.558, ' The label is': 0.56}},
  'premise: ': {' ': {'': 0.534, ' The label is': 0.538},
   ' Consider the following hypothesis: ': {'': 0.598, ' The label is': 0.598},
   ' The following hypothesis is true: ': {'': 0.602, ' The label is': 0.588},
   ' hypothesis: ': {'': 0.594, ' The label is': 0.588}}}}

max = 0.61
min = 0.14#bzw 0
#scheint übereinzustimmen mit 0:500

In [ ]:
# mnli,0:500 validation matched


{'': {' ': {' ': {'': 0.0, ' The label is': 0.0},
   ' Consider the following hypothesis: ': {'': 0.618, ' The label is': 0.6},
   ' The following hypothesis is true: ': {'': 0.152, ' The label is': 0.126},
   ' hypothesis: ': {'': 0.592, ' The label is': 0.594}},
  'Consider the following premise: ': {' ': {'': 0.436,
    ' The label is': 0.412},
   ' Consider the following hypothesis: ': {'': 0.598, ' The label is': 0.57},
   ' The following hypothesis is true: ': {'': 0.504, ' The label is': 0.49},
   ' hypothesis: ': {'': 0.596, ' The label is': 0.59}},
  'The following premise is true: ': {' ': {'': 0.162, ' The label is': 0.14},
   ' Consider the following hypothesis: ': {'': 0.59, ' The label is': 0.578},
   ' The following hypothesis is true: ': {'': 0.576, ' The label is': 0.534},
   ' hypothesis: ': {'': 0.582, ' The label is': 0.576}},
  'premise: ': {' ': {'': 0.566, ' The label is': 0.552},
   ' Consider the following hypothesis: ': {'': 0.62, ' The label is': 0.63},
   ' The following hypothesis is true: ': {'': 0.596, ' The label is': 0.594},
   ' hypothesis: ': {'': 0.596, ' The label is': 0.606}}},
 'mnli ': {' ': {' ': {'': 0.0, ' The label is': 0.0},
   ' Consider the following hypothesis: ': {'': 0.604, ' The label is': 0.616},
   ' The following hypothesis is true: ': {'': 0.388, ' The label is': 0.342},
   ' hypothesis: ': {'': 0.58, ' The label is': 0.574}},
  'Consider the following premise: ': {' ': {'': 0.39, ' The label is': 0.38},
   ' Consider the following hypothesis: ': {'': 0.596, ' The label is': 0.588},
   ' The following hypothesis is true: ': {'': 0.56, ' The label is': 0.546},
   ' hypothesis: ': {'': 0.564, ' The label is': 0.58}},
  'The following premise is true: ': {' ': {'': 0.36, ' The label is': 0.348},
   ' Consider the following hypothesis: ': {'': 0.538, ' The label is': 0.514},
   ' The following hypothesis is true: ': {'': 0.558, ' The label is': 0.554},
   ' hypothesis: ': {'': 0.552, ' The label is': 0.558}},
  'premise: ': {' ': {'': 0.54, ' The label is': 0.53},
   ' Consider the following hypothesis: ': {'': 0.582, ' The label is': 0.588},
   ' The following hypothesis is true: ': {'': 0.584, ' The label is': 0.588},
   ' hypothesis: ': {'': 0.584, ' The label is': 0.582}}}}

max = 0.62
min = 0.12#bzw 0
# auch hier: Z ist größtenteils egal, teh label is stört eher

In [ ]:
# snli,0:500 test

{'': {' ': {' ': {'': 0.0, ' The label is': 0.0},
   ' Consider the following hypothesis: ': {'': 0.458, ' The label is': 0.458},
   ' The following hypothesis is true: ': {'': 0.226, ' The label is': 0.196},
   ' hypothesis: ': {'': 0.46, ' The label is': 0.446}},
  'Consider the following premise: ': {' ': {'': 0.38, ' The label is': 0.362},
   ' Consider the following hypothesis: ': {'': 0.456, ' The label is': 0.442},
   ' The following hypothesis is true: ': {'': 0.39, ' The label is': 0.372},
   ' hypothesis: ': {'': 0.462, ' The label is': 0.456}},
  'The following premise is true: ': {' ': {'': 0.164, ' The label is': 0.12},
   ' Consider the following hypothesis: ': {'': 0.458, ' The label is': 0.432},
   ' The following hypothesis is true: ': {'': 0.454, ' The label is': 0.452},
   ' hypothesis: ': {'': 0.478, ' The label is': 0.456}},
  'premise: ': {' ': {'': 0.402, ' The label is': 0.392},
   ' Consider the following hypothesis: ': {'': 0.48, ' The label is': 0.462},
   ' The following hypothesis is true: ': {'': 0.472, ' The label is': 0.46},
   ' hypothesis: ': {'': 0.472, ' The label is': 0.474}}},
 'mnli ': {' ': {' ': {'': 0.0, ' The label is': 0.0},
   ' Consider the following hypothesis: ': {'': 0.446, ' The label is': 0.442},
   ' The following hypothesis is true: ': {'': 0.386, ' The label is': 0.376},
   ' hypothesis: ': {'': 0.48, ' The label is': 0.468}},
  'Consider the following premise: ': {' ': {'': 0.342,
    ' The label is': 0.348},
   ' Consider the following hypothesis: ': {'': 0.47, ' The label is': 0.446},
   ' The following hypothesis is true: ': {'': 0.446, ' The label is': 0.432},
   ' hypothesis: ': {'': 0.484, ' The label is': 0.462}},
  'The following premise is true: ': {' ': {'': 0.328, ' The label is': 0.322},
   ' Consider the following hypothesis: ': {'': 0.468, ' The label is': 0.45},
   ' The following hypothesis is true: ': {'': 0.466, ' The label is': 0.444},
   ' hypothesis: ': {'': 0.496, ' The label is': 0.484}},
  'premise: ': {' ': {'': 0.438, ' The label is': 0.426},
   ' Consider the following hypothesis: ': {'': 0.468, ' The label is': 0.462},
   ' The following hypothesis is true: ': {'': 0.472, ' The label is': 0.47},
   ' hypothesis: ': {'': 0.472, ' The label is': 0.462}}}}

max = 0.468
min = 0.12#bzw 0
# "the label is" stört eher, hat aber geringen einfluss






*   mnli als triggerwort stört eher
*   hypothesis: muss als trigger drin stehen
*   reihenfolge ist entscheidend. Wenn man premise und hypothesis swapped ist acc viel niedriger
*   beam search erhöht acc
*   hohe acc bei snli, obowohl es darauf(wahrscheinlich) nicht trainiert wurde
*   t5 zuerst die prämisse zusammenfassen lassen (summarize: ) führt zu keinen acc verlusten(z.T. leichte acc gains)(kann sein, dass t5 bei kurzen prämissen einfach nichts ändert)
*   t5 zuerste in german/french/romanian übersetzen lassen sieht verluste in acc, aber immer noch weit über random(60%)
*   auch translation mit anderem transformer(mBart) funktioniert. für deutsch sogar knapp 10% mehr acc(69 statt 60). Auch spanish translation weit über random acc(50). Übersetzen in Koreanisch oder Schwedisch hingegen führt zu random acc(spanisch sehr ähnlich zu französisch?)
*   cb statt mnli kein Unterschied
*   nur prämisse oder nur hypothese führt zu random acc
*   in mrpc übersetzen(task bei dem man bestimmen soll, ob zwei sätze äquivalent sind) führt zu acc von 47, also mehr als random, wenn man "equivalent" = "entailment" und "non_equivalent"= "contradiction " macht. Es ist zu beachten, dass es hier t5 garnicht möglich ist neutral auszugeben, also maximale acc ~70% (es kommt zudem auf die reihenfolge an, zuerst prämisse, dann hypothese ist fast random acc)
*   in qnli übersetzen auch mehr als random acc, obowhl nie neutral ausgegeben werden kann(aufgabe ist es zu bestimmen, ob eine gegebene frage aus einem kontext beantwortbar ist, hier einfach hypothese als frage und premise als kontext)
*   {'question: Should the label be entailment, contradiction or neutral? context: ': {'hypothesis: ': {' premise: ': {'': 0.5}}}}(squad format), über random acc, aber man muss trigger im kontext lassen, sonst muss man antwort anders bestimmen
*   Einzelne vocale und mehrere vocale zufällig durch andere zu ersetzen hat immer noch mehr als random acc
*  t5 schaut auf sentiment, auch in fällen wo dies kein sinn ergibt
*  t5 achtet auf token overlapp
*  Reihenfolge der token ist egal?
*  hyp und premise vertauscht hat auch acc weit über random(55%), kann natürlich sein dass es im datensatz auch einige äquivalenzen und nicht nur implikationen gibt
